### ISSUE TITLE
False-positive "Name '...' already defined

### ISSUE THREAD

**Issue by bersbersbers** [python/mypy#10358]
**Bug Report**

```python
"""False-positive "Name 'x' already defined on line 3."""
if 1 < 0:
    x = 0
else:
    [(x := 0) for _ in range(1)]
print(x)
```

**To Reproduce**
`mypy bug.py`

**Expected Behavior**

No error.

**Actual Behavior**

```
../bug.py:5: error: Name 'x' already defined on line 3
Found 1 error in 1 file (checked 1 source file)
```
**Your Environment**

- Mypy version used: 0.812
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.8.9
- Operating system and version: Linux
